content/2016-04-18-aliasing-improved.rst

```diff
@@ -0,0 +1,473 @@
+is more complex:
+
+.. code:: python
+
+    >>> module = ast.parse('def foo(a, b): return a or b[0]')
+    >>> result = pm.gather(Aliases, module)
+    >>> f = module.body[0].return_alias
+    >>> List = ast.List([ast.Name('L0', ast.Load())], ast.Load())
+    >>> Aliases.dump(f([ast.Name('B', ast.Load()), List]))
+    ['B', '[L0][0]']
+
+Which actually means that when called with two arguments ``B`` and
+the single-element list ``[L[0]]``, ``foo`` may returns either the
+first argument, or the first element of the second argument.
+
+
+Assign Statement
+----------------
+
+Assignment creates aliasing between lhs and rhs
+
+.. code:: python
+
+    >>> module = ast.parse('def foo(a): c = a ; d = e = c ; {c, d, e}')
+    >>> result = pm.gather(Aliases, module)
+    >>> Aliases.dump(result, filter=ast.Set)
+    {c, d, e} => ['|a|', '|a|', '|a|']
+
+Everyone points to the formal parameter ``a`` \o/
+
+For Statement
+-------------
+
+For loop creates aliasing between the target
+and the content of the iterator
+
+.. code:: python
+
+    >>> module = ast.parse('def foo(a):\n for i in a:\n  {i}')
+    >>> result = pm.gather(Aliases, module)
+    >>> Aliases.dump(result, filter=ast.Set)
+    {i} => ['|i|']
+
+Not very useful, unless we know something about the iterated container
+
+.. code:: python
+
+    >>> module = ast.parse('def foo(a, b):\n for i in [a, b]:\n  {i}')
+    >>> result = pm.gather(Aliases, module)
+    >>> Aliases.dump(result, filter=ast.Set)
+    {i} => ['|a|', '|b|']
+
+If Statement
+------------
+
+After an if statement, the values from both branches are merged,
+potentially creating more aliasing:
+
+.. code:: python
+
+    >>> fun = 'def foo(a, b):\n if a: c=a\n else: c=b\n return {c}'
+    >>> module = ast.parse(fun)
+    >>> result = pm.gather(Aliases, module)
+    >>> Aliases.dump(result, filter=ast.Set)
+    {c} => ['|a|', '|b|']
+
+Illustration: Typing
+--------------------
+
+Thanks to the above analyse, Pythran is capable of computing some rather difficult informations! In the following:
+
+.. code:: python
+
+    def typing_aliasing_and_variable_subscript_combiner(i):
+        a=[list.append,
+           lambda x,y: x.extend([y])
+        ]
+        b = []
+        a[i](b, i)
+        return b
+
+Pythran knows that ``b`` is a list of elements of the same type as ``i``.
+
+And in the following:
+
+.. code:: python
+
+    def typing_and_function_dict(a):
+        funcs = {
+            'zero' : lambda x: x.add(0),
+            'one' : lambda x: x.add(1),
+        }
+        s = set()
+        funcs[a](s)
+        return s
+
+Pythran knows that ``s`` is a set of integers :-)
+
+
+.. [0] Except the sanity of the developer, but who never used the ``id`` or ``len`` identifiers?
+
+.. [1] Starting from this note, the identifiers from the `ast <https://docs.python.org/2/library/ast.html>`_ module are used.
+
+.. [git-version] The Pythran commit id used for this article is 0000eeeee
```

Comment:
> Good id?

Still changing as I integrate some of your remark on the docstring in
the original code :-)